examples/auction/src/lib.rs

```diff
@@ -166,35 +177,24 @@ fn auction_finalize<S: HasStateApi>(
     let slot_time = ctx.metadata().slot_time();
     ensure!(slot_time > state.viewable_state.expiry, FinalizeError::AuctionStillActive);
 
-    let owner = ctx.owner();
-
     let balance = host.self_balance();
     if balance == Amount::zero() {
         Ok(())
     } else {
-        if host.invoke_transfer(&owner, state.viewable_state.highest_bid).is_err() {
+        let owner = ctx.owner();
+
+        // Sends the highest bid to the owner
+        if host.invoke_transfer(&owner, balance).is_err() {
             bail!(FinalizeError::BidMapError);
-        }
-        let mut remaining_bid = None;
-        // Return bids that are smaller than highest
-        for (addr, amnt) in state.bids.iter() {
-            if *amnt < state.viewable_state.highest_bid {
-                if host.invoke_transfer(&*addr, *amnt).is_err() {
-                    bail!(FinalizeError::BidMapError);
+        } else {
+            match state.viewable_state.highest_bidder {
+                Some(account_address) => {
+                    host.state_mut().viewable_state.auction_state =
+                        AuctionState::Sold(account_address);
+                    Ok(())
```

Comment:
I would restructure this into something like
```rust
if let Some(account_address) = state.viewable_state.highest_bidder.take() {
    host.state_mut().viewable_state.auction_state = AuctionState::Sold(account_address);
    if host.invoke_transfer(&owner, balance).is_err() {
        ...
    }
}
```
so that you don't attempt a transfer when no bids are made. 

At the same time you can also get rid of the outer `if balance == zero` check since if there is a bid the amount must be non-zero.

Reply:
addressed